src/prof_recent.c

```diff
@@ -466,67 +479,95 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
-	emitter_begin(&emitter);
 
-	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
-	prof_recent_alloc_assert_count(tsd);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
+	emitter_begin(&emitter);
 
 	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
+	 * Temporarily set prof_recent_alloc_max to -1 so that no existing
+	 * record would be lost during dumping.
 	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	ssize_t max = prof_recent_alloc_max_write(tsd, -1);
 
 	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
-
 	emitter_json_array_kv_begin(&emitter, "recent_alloc");
-	for (prof_recent_t *n = prof_recent_alloc_begin(tsd);
-	    n != prof_recent_alloc_end(tsd);
-	    n = prof_recent_alloc_next(tsd, n)) {
-		emitter_json_object_begin(&emitter);
-
-		emitter_json_kv(&emitter, "usize", emitter_type_size,
-		    &n->usize);
-		bool released = n->alloc_edata == NULL;
-		emitter_json_kv(&emitter, "released", emitter_type_bool,
-		    &released);
-
-		emitter_json_kv(&emitter, "alloc_thread_uid",
-		    emitter_type_uint64, &n->alloc_tctx->thr_uid);
-		uint64_t alloc_time_ns = nstime_ns(&n->alloc_time);
-		emitter_json_kv(&emitter, "alloc_time", emitter_type_uint64,
-		    &alloc_time_ns);
-		emitter_json_array_kv_begin(&emitter, "alloc_trace");
-		dump_bt(&emitter, n->alloc_tctx);
-		emitter_json_array_end(&emitter);
-
-		if (n->dalloc_tctx != NULL) {
-			assert(released);
-			emitter_json_kv(&emitter, "dalloc_thread_uid",
-			    emitter_type_uint64, &n->dalloc_tctx->thr_uid);
-			assert(!nstime_equals_zero(&n->dalloc_time));
-			uint64_t dalloc_time_ns = nstime_ns(&n->dalloc_time);
-			emitter_json_kv(&emitter, "dalloc_time",
-			    emitter_type_uint64, &dalloc_time_ns);
-			emitter_json_array_kv_begin(&emitter, "dalloc_trace");
-			dump_bt(&emitter, n->dalloc_tctx);
+
+	prof_recent_t *n = NULL;
+	ssize_t accum = 0;
+	while (true) {
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		prof_recent_alloc_assert_count(tsd);
+
+		if (n == NULL) {
+			n = prof_recent_alloc_begin(tsd);
+		}
+
+		bool need_more_batch = false;
+		for (int count = 0; accum < max &&
+		    n != prof_recent_alloc_end(tsd); ++count) {
+			if (count == PROF_RECENT_PRINT_BATCH) {
+				need_more_batch = true;
+				break;
+			}
+
+			emitter_json_object_begin(&emitter);
+
+			emitter_json_kv(&emitter, "usize",
+			    emitter_type_size, &n->usize);
+			bool released = n->alloc_edata == NULL;
+			emitter_json_kv(&emitter, "released",
+			    emitter_type_bool, &released);
+
+			emitter_json_kv(&emitter, "alloc_thread_uid",
+			    emitter_type_uint64, &n->alloc_tctx->thr_uid);
+			uint64_t alloc_time_ns = nstime_ns(&n->alloc_time);
+			emitter_json_kv(&emitter, "alloc_time",
+			    emitter_type_uint64, &alloc_time_ns);
+			emitter_json_array_kv_begin(&emitter, "alloc_trace");
+			dump_bt(&emitter, n->alloc_tctx);
```

Comment:
It would be better to not do the actual dumping holding the recent lock. Even with a small batch size, if the dump involves IO there will be latency spikes for regular requests. Let’s try read out the information under the recent lock (keep the info on stack), and drop the lock when dump?

Also thanks for getting to this quickly!

Reply:
Do you first read out everything under the lock, and then drop the lock and write it out? We may need a significant amount of space to store all the information. One allocation record can output up to 800-900 characters.

Reply:
I see what you mean. Changed to: read out each batch under the lock, drop the lock and do I/O. Since the buffer is large enough to hold the entire batch, there should not be any I/O during the reading. (It's hard to make an absolute guarantee because there's no way to accurately predict the output size of the batch upfront.)

Reply:
I'd say to push it a bit further, and guarantee non-blocking: we can add a `_nonblocking` version of the dump_bt, which attempts only append to the buffer w/o actually dumping, and returns whether it could do so. That way we don't really need a batch size, we can simply keep buffering the output, and only unlock and dump when the buffer is full.

Reply:
Discussed offline, will develop a different buffered writer to meet such needs.